File: runtime-rs/tests/golden.rs
//! Тест: компилируемый рантайм совпадает с эталонными
//! векторами, сгенерированными Python-реализацией (mayak/runtime/golden.py).
//!
//! Эталон лежит в tests/data/runtime_golden репозитория. Пересоздание:
//!     python scripts/make_runtime_golden.py
use std::path::PathBuf;

use mayak_rt::calendar::doy_hour;
use mayak_rt::calib::{aci_score, apply_adaptive, apply_conformal, AciParams};
use mayak_rt::state::{Snapshot, HEADER_V3};
use mayak_rt::{Manifest, Runtime, RuntimeOptions};
use serde_json::Value;

fn golden_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../tests/data/runtime_golden")
}

struct Golden {
    doc: Value,
    blob: Vec<f32>,
}

fn load() -> Golden {
    let dir = golden_dir();
    let doc: Value = serde_json::from_str(&std::fs::read_to_string(dir.join("golden.json")).unwrap()).unwrap();
    let raw = std::fs::read(dir.join("golden.f32")).unwrap();
    let blob = raw
        .as_chunks::<4>()
        .0
        .iter()
        .map(|b| f32::from_le_bytes([b[0], b[1], b[2], b[3]]))
        .collect();
    Golden { doc, blob }
}

impl Golden {
    fn take(&self, r: &Value) -> &[f32] {
        let (o, n) = (
            r["offset"].as_u64().unwrap() as usize,
            r["len"].as_u64().unwrap() as usize,
        );
        &self.blob[o..o + n]
    }
}

fn f(v: &Value) -> f32 {
    v.as_f64().unwrap() as f32
}

fn obs(v: &Value) -> Option<f64> {
    match v {
        Value::Null => None,
        Value::String(s) if s == "nan" => Some(f64::NAN),
        v => v.as_f64(),
    }
}

fn score(v: &Value) -> f64 {
    match v.as_str() {
        Some("nan") => f64::NAN,
        Some("inf") => f64::INFINITY,
        _ => v.as_f64().unwrap(),
    }
}

fn max_abs(a: &[f32], b: &[f32]) -> f32 {
    assert_eq!(a.len(), b.len());
    a.iter().zip(b).map(|(x, y)| (x - y).abs()).fold(0.0, f32::max)
}

fn runtime(sc: &Value) -> Runtime {
    let opts = RuntimeOptions {
        conformal: sc["conformal"].as_bool().unwrap(),
        aci: sc["aci"].as_bool().unwrap(),
        ..Default::default()
    };
    Runtime::new(
        golden_dir().join("model"),
        sc["lat"].as_f64().unwrap(),
        sc["lon"].as_f64().unwrap(),
        sc["elev"].as_f64().unwrap(),
        &opts,
    )
    .unwrap()
}

fn scenario(g: &Golden, name: &str) -> &'static Value {
    let sc = g.doc["scenarios"]
        .as_array()
        .unwrap()
        .iter()
        .find(|s| s["name"] == name)
        .unwrap()
        .clone();
    Box::leak(Box::new(sc))
}

/// Прогон сценария → max|Δq| по записанным выпускам.
fn replay(g: &Golden, sc: &Value, rt: &mut Runtime) -> f32 {
    let year = g.doc["year_hours"].as_i64().unwrap();
    let tol = g.doc["tolerance"]["q_abs"].as_f64().unwrap() as f32;
    let (h, nq) = (rt.horizon(), rt.n_quantiles());
    let (mut doy, mut hour) = (vec![0.0f32; h], vec![0.0f32; h]);
    let mut worst = 0.0f32;
    for (i, ev) in sc["events"].as_array().unwrap().iter().enumerate() {
        if ev["op"] == "step" {
            let o = ev["obs"].as_array().unwrap();
            rt.step([obs(&o[0]), obs(&o[1]), obs(&o[2])], f(&ev["doy"]), f(&ev["hour"]))
                .unwrap();
            continue;
        }
        let h0 = ev["fut_hoy0"].as_i64().unwrap();
        for k in 0..h {
            let hoy = (h0 + k as i64).rem_euclid(year);
            doy[k] = (hoy as f64 / 24.0) as f32;
            hour[k] = (hoy % 24) as f32;
        }
        let fc = rt.forecast(&doy, &hour).unwrap();
        assert!(!fc.fallback);
        if !ev["q"].is_null() {
            let refq = g.take(&ev["q"]);
            assert_eq!(refq.len(), h * nq);
            let err = max_abs(&fc.q, refq);
            assert!(
                err <= tol,
                "{}: событие {i}: max|Δq| = {err:.2e} > {tol:.0e}",
                sc["name"]
            );
            for row in fc.q.chunks(nq) {
                assert!(row.windows(2).all(|w| w[0] <= w[1]), "квантили не монотонны");
            }
            worst = worst.max(err);
        }
        let dt = (rt.theta() - f(&ev["theta"])).abs();
        assert!(
            dt <= 1e-6,
            "{}: событие {i}: θ = {} против эталона {}",
            sc["name"],
            rt.theta(),
            ev["theta"]
        );
    }
    let fin = &sc["final"];
    assert!((rt.theta() - f(&fin["theta"])).abs() <= 1e-6);
    assert_eq!(
        rt.aci_updates(),
        fin["aci_updates"].as_u64().unwrap(),
        "{}: число обратных связей ACI",
        sc["name"]
    );
    assert_eq!(
        rt.aci_misses(),
        fin["aci_misses"].as_u64().unwrap(),
        "{}: число промахов ACI",
        sc["name"]
    );
    assert_eq!(rt.calendar_breaks(), fin["calendar_breaks"].as_u64().unwrap());
    assert_eq!(rt.filled() as u64, fin["filled"].as_u64().unwrap());
    assert_eq!(rt.hours_in_day() as u64, fin["hours_in_day"].as_u64().unwrap());
    println!("{}: max|Δq| Rust ↔ эталон = {worst:.3e} °C", sc["name"]);
    worst
}

#[test]
fn cold_start_with_conformal_and_aci() {
    let g = load();
    let sc = scenario(&g, "cold_aci");
    replay(&g, sc, &mut runtime(sc));
}

#[test]
fn extremes_edges_gaps_calendar_break() {
    let g = load();
    let sc = scenario(&g, "extremes");
    replay(&g, sc, &mut runtime(sc));
}

#[test]
fn restart_from_python_state_restores_incomplete_day() {
    let g = load();
    let sc = scenario(&g, "restart");
    let raw = std::fs::read(golden_dir().join(sc["init_state"].as_str().unwrap())).unwrap();
    let mut rt = runtime(sc);
    rt.load_state(&raw).unwrap();
    assert!(rt.hours_in_day() > 0, "эталон должен начинаться с незавершённых суток");
    // Состояние Python → Rust → байты: совпадение до байта (формат v3 общий).
    let mut back = Vec::new();
    rt.serialize(&mut back);
    assert_eq!(back, raw, "serialize(load(S)) ≠ S");
    replay(&g, sc, &mut rt);

    // Конечное состояние: курсор, календарь, окно и маска - точно; числа - в допуске.
    let end = std::fs::read(golden_dir().join(sc["final"]["state"].as_str().unwrap())).unwrap();
    rt.serialize(&mut back);
    assert_eq!(back.len(), end.len());
    assert_eq!(back[..12], end[..12], "заголовок состояния");
    let d = &rt.manifest.dims;
    let tail = d.stream_window * 3 * 3;
    assert_eq!(back[back.len() - tail..], end[end.len() - tail..], "сырое окно и маска");
    let (a, b) = (Snapshot::parse(&back, d).unwrap(), Snapshot::parse(&end, d).unwrap());
    for (x, y) in [
        (&a.n_re, &b.n_re),
        (&a.n_im, &b.n_im),
        (&a.e, &b.e),
        (&a.z, &b.z),
        (&a.day_summ, &b.day_summ),
    ] {
        let scale = y.iter().fold(1.0f32, |m, v| m.max(v.abs()));
        assert!(max_abs(x, y) <= 1e-3 * scale, "числовая часть состояния расходится");
    }
    assert_eq!(a.day_mask, b.day_mask);
    assert!((a.theta - b.theta).abs() <= 1e-6);
    assert_eq!(back.len(), HEADER_V3 + (end.len() - HEADER_V3));
}

#[test]
fn state_size_is_pinned() {
    let m = Manifest::load(golden_dir().join("model")).unwrap();
    assert_eq!(mayak_rt::state::nbytes(&m.dims), 3352);
    assert_eq!(m.state.nbytes, 3352);
}

#[test]
fn corrupted_state_is_rejected_and_leaves_cold_start() {
    let g = load();
    let sc = scenario(&g, "restart");
    let raw = std::fs::read(golden_dir().join(sc["init_state"].as_str().unwrap())).unwrap();
    let mut rt = runtime(sc);
    for bad in [&raw[..raw.len() - 1], &[b'X'; 3352][..]] {
        assert!(rt.load_state(bad).is_err());
    }
    let mut v = raw.clone();
    v[6] = 30; // часов в сутках > 23
    assert!(rt.load_state(&v).is_err());
    assert_eq!(rt.filled(), 0);
}

#[test]
fn calendar_matches_timeaxis() {
    let g = load();
    let c = &g.doc["calendar"];
    let hours = c["unix_hours"].as_array().unwrap();
    for (i, h) in hours.iter().enumerate() {
        let (d, hr) = doy_hour(h.as_i64().unwrap());
        assert_eq!(d.to_bits(), f(&c["doy"][i]).to_bits(), "doy часа {h}");
        assert_eq!(hr.to_bits(), f(&c["hour"][i]).to_bits(), "hour часа {h}");
    }
}

#[test]
fn calibration_matches_metrics() {
    let g = load();
    let m = Manifest::load(golden_dir().join("model")).unwrap();
    let table = m.conformal_table().unwrap().unwrap();
    let (nq, im) = (m.dims.n_quantiles, m.i_med);
    let cal = &g.doc["calibration"];
    for case in cal["cases"].as_array().unwrap() {
        let mut q = g.take(&case["q"]).to_vec();
        if case["conformal"].as_bool().unwrap() {
            apply_conformal(&mut q, &table, nq);
        }
        apply_adaptive(&mut q, f(&case["theta"]), nq, im);
        let want = g.take(&case["expect"]);
        for (x, y) in q.iter().zip(want) {
            assert_eq!(x.to_bits(), y.to_bits(), "калибровка расходится до бита");
        }
    }
    let a = m.calibration.aci.as_ref().unwrap();
    let p = AciParams {
        target: a.target,
        gamma: a.gamma,
        max_factor: a.max_factor,
        interval: a.interval,
    };
    let aci = &cal["aci"];
    let mut theta = p.clip(aci["theta0"].as_f64().unwrap());
    for (k, s) in aci["scores"].as_array().unwrap().iter().enumerate() {
        assert_eq!(theta, f(&aci["theta_before"][k]), "θ перед наблюдением {k}");
        let s = score(s);
        if s.is_nan() {
            continue;
        }
        let (t, miss) = p.step(theta, s);
        assert_eq!(miss, aci["miss"][k].as_bool().unwrap());
        theta = t;
    }
    assert_eq!(theta, f(&aci["theta_end"]));
    let sc = &cal["score"];
    let qs = g.take(&sc["q"]);
    for (k, y) in sc["y"].as_array().unwrap().iter().enumerate() {
        let got = aci_score(f(y) as f64, &qs[k * nq..(k + 1) * nq], p.interval, im);
        let want = score(&sc["expect"][k]);
        assert!(
            got == want || (got.is_nan() && want.is_nan()),
            "aci_score {k}: {got} против {want}"
        );
    }
}

File: mayak/runtime/golden.py
"""Эталонные векторы компилируемого рантайма.

Эталон - Python: потоковый рантайм ``StreamingMayak``, калибровка
``mayak.metrics``, календарь ``mayak.timeaxis``. Векторы
генерируются на детерминированной модели (сид, все параметры «встряхнуты», чтобы нулевые
инициализации голов, FiLM и подстройки мод не прятали ошибки), записываются в
tests/data/runtime_golden/ и коммитятся. Рантайм на Rust (runtime-rs/tests/golden.rs)
сверяется с ними.

Состав:
* model/                 - четыре графа ONNX, manifest.json, conformal.f32;
* golden.json            - сценарии (события step/forecast), календарь, калибровка;
* golden.f32             - ожидаемые квантили и входы калибровки (float32 LE);
* state_*.bin            - состояния, записанные эталоном (формат v3).

Сценарии:
* cold_aci  - холодный старт, конформная таблица и ACI, переход через Новый год,
              дробные наблюдения, которые хост записывает целыми градусами и
              процентами, значения рядом с границами физических диапазонов до и после
              записи, половины между целыми, NaN, пропуски;
* restart   - старт из состояния эталона с незавершёнными сутками, ACI с θ ≠ 0;
* extremes  - полярная точка, 30 ч без данных, значения ровно на границах диапазонов,
              разрыв календаря, без калибровки.
"""
from __future__ import annotations

import json
import os

import numpy as np
import torch

from mayak.metrics import ACIParams, aci_run, aci_score, calibrate_forecast
from mayak.runtime.equivalence import synthetic_series
from mayak.runtime.streaming import StreamingMayak
from mayak.timeaxis import window_calendar

GOLDEN_FORMAT = 1
GOLDEN_SEED = 1414
GOLDEN_PERTURB = 0.05
YEAR_H = 365 * 24
GOLDEN_ACI = ACIParams(target=0.10, gamma=0.05, max_factor=4.0)
GOLDEN_SHIFT = (np.array([-0.3, -0.2, -0.08, 0.0, 0.08, 0.2, 0.3], np.float32)[None, :]
                * np.array([1.0, 1.5, 2.0, 2.5], np.float32)[:, None]
                + np.array([0.0, 0.0, 0.0, 0.02, 0.05, 0.05, 0.1], np.float32))
MIN_ACI_MARGIN = 1e-3
Q_ATOL = 5e-4
FRESH_ATOL = 2e-4
DEFAULT_DIR = os.path.join("tests", "data", "runtime_golden")


def golden_model(cfg=None):
    """Детерминированная модель эталона: сид + шум на всех параметрах."""
    from mayak.model import MAYAK
    torch.manual_seed(GOLDEN_SEED)
    m = MAYAK(cfg).eval()
    with torch.no_grad():
        for p in m.parameters():
            p.add_(GOLDEN_PERTURB * torch.randn_like(p))
    return m


def _f(x):
    """float32 → float Python (repr JSON однозначно возвращает то же float32)."""
    return float(np.float32(x))


def _enc_obs(v):
    """Наблюдение → JSON: число, null (нет данных) или "nan" (NaN от датчика)."""
    if v is None:
        return None
    return "nan" if v != v else float(v)


def dec_obs(v):
    return float("nan") if v == "nan" else v


def _enc_score(v):
    return "nan" if np.isnan(v) else ("inf" if np.isinf(v) else float(v))


def _obs(series, k):
    return [_f(series["x"][k, j]) if series["m"][k, j] > 0 else None for j in range(3)]


def _fut(hoy0, horizon):
    hoy = (hoy0 + np.arange(horizon)) % YEAR_H
    return (hoy / 24.0).astype(np.float32), (hoy % 24).astype(np.float32)


def _cal(hoy):
    hoy = int(hoy) % YEAR_H
    return _f(hoy / 24.0), _f(hoy % 24)


class _Recorder:
    """Прогон эталона по событиям с записью ожидаемых значений."""

    def __init__(self, stream, blob):
        self.s, self.blob = stream, blob
        self.events = []
        self.margins = []

    def step(self, obs, doy, hour):
        s = self.s
        if s.aci is not None and s._pending is not None and obs[0] is not None:
            p = s._pending
            hit = np.flatnonzero(p["hoy"] == int(round(float(np.float32(doy)) * 24.0)))
            if hit.size and int(hit[0]) > p["last"]:
                sc = float(aci_score(float(np.float32(obs[0])), p["q"][int(hit[0])],
                                     s.aci.interval))
                if np.isfinite(sc):
                    self.margins.append(abs(sc - np.exp(s.theta)))
        s.step(*obs, np.float32(doy), np.float32(hour))
        self.events.append(dict(op="step", obs=[_enc_obs(v) for v in obs], doy=_f(doy),
                                hour=_f(hour)))

    def forecast(self, hoy0, record=True):
        """Выпуск; record=False - выпуск нужен только как обратная связь ACI, квантили
        не пишутся (объём эталона), θ пишется всегда."""
        doy, hour = _fut(hoy0, self.s.m.cfg.horizon)
        q, _mu = self.s.forecast(doy, hour)
        self.events.append(dict(op="forecast", fut_hoy0=int(hoy0),
                                q=self.blob.put(q) if record else None,
                                theta=_f(self.s.theta)))

    def final(self):
        s = self.s
        return dict(theta=_f(s.theta), aci_updates=s.aci_updates, aci_misses=s.aci_misses,
                    calendar_breaks=s.calendar_breaks, filled=s.filled,
                    hours_in_day=s._hours_in_day)


class _Blob:
    def __init__(self):
        self.parts, self.n = [], 0

    def put(self, a):
        a = np.ascontiguousarray(a, "<f4").ravel()
        off = self.n
        self.parts.append(a)
        self.n += a.size
        return dict(offset=off, len=int(a.size))

    def array(self):
        return np.concatenate(self.parts) if self.parts else np.zeros(0, "<f4")


def _series(n, seed, start_hoy):
    s = synthetic_series(n, seed=seed, start_hoy=start_hoy)
    s["hoy"] = (start_hoy + np.arange(n)) % YEAR_H
    return s


def scenario_cold_aci(model, blob):
    lat, lon, elev = 52.37, 4.9, -2.0
    st = StreamingMayak(model, lat, lon, elev, conformal=GOLDEN_SHIFT, aci=GOLDEN_ACI)
    rec = _Recorder(st, blob)
    H = model.cfg.horizon
    s = _series(330, seed=11, start_hoy=YEAR_H - 100)
    bad = {40: (75.0, None, None), 41: (None, None, -5.0), 42: (None, 250.0, None),
           43: (float("nan"), None, None), 44: (60.6, 1100.06, 100.6),
           45: (60.4, 1100.04, 100.4), 46: (-0.5, 1013.25, 12.5)}
    rec.forecast(s["hoy"][0])
    for k in range(len(s["hoy"])):
        obs = _obs(s, k)
        if k in bad:
            obs = [b if b is not None else o for b, o in zip(bad[k], obs)]
        rec.step(obs, *_cal(s["hoy"][k]))
        if k % 6 == 5:
            rec.forecast(s["hoy"][k] + 1, record=k % 24 == 23)
    return dict(name="cold_aci", lat=lat, lon=lon, elev=elev, conformal=True, aci=True,
                init_state=None, events=rec.events, final=rec.final()), rec.margins, H


def scenario_restart(model, blob, out_dir):
    lat, lon, elev = -33.87, 151.21, 58.0
    s = _series(300 + 96, seed=23, start_hoy=2000)
    a = StreamingMayak(model, lat, lon, elev, conformal=GOLDEN_SHIFT, aci=GOLDEN_ACI)
    pre = _Recorder(a, _Blob())
    for k in range(300):
        pre.step(_obs(s, k), *_cal(s["hoy"][k]))
        if k % 8 == 7:
            pre.forecast(s["hoy"][k] + 1)
    raw = a.serialize()
    with open(os.path.join(out_dir, "state_restart.bin"), "wb") as fh:
        fh.write(raw)
    b = StreamingMayak(model, lat, lon, elev, conformal=GOLDEN_SHIFT, aci=GOLDEN_ACI)
    b.load_state(raw)
    rec = _Recorder(b, blob)
    rec.forecast(s["hoy"][299] + 1)
    for k in range(300, len(s["hoy"])):
        rec.step(_obs(s, k), *_cal(s["hoy"][k]))
        if k % 12 == 11:
            rec.forecast(s["hoy"][k] + 1)
    end = b.serialize()
    with open(os.path.join(out_dir, "state_restart_end.bin"), "wb") as fh:
        fh.write(end)
    fin = rec.final()
    fin["state"] = "state_restart_end.bin"
    return dict(name="restart", lat=lat, lon=lon, elev=elev, conformal=True, aci=True,
                init_state="state_restart.bin", events=rec.events,
                final=fin), pre.margins + rec.margins


def scenario_extremes(model, blob):
    lat, lon, elev = 78.22, 15.65, 2000.0
    st = StreamingMayak(model, lat, lon, elev)
    rec = _Recorder(st, blob)
    hoy = 4300
    for _ in range(30):
        rec.step([None, None, None], *_cal(hoy))
        hoy += 1
    rec.forecast(hoy)
    edges = [(-90.0, 300.0, 0.0), (60.0, 1100.0, 100.0), (-90.0, 1100.0, 100.0),
             (60.0, 300.0, 0.0)]
    for k in range(60):
        rec.step(list(edges[k % 4]), *_cal(hoy))
        hoy += 1 if k != 30 else 2                        # разрыв календаря на час
    rec.forecast(hoy)
    rec.step([-5.0, 800.0, 55.0], *_cal(hoy))
    rec.forecast(hoy + 1)
    return dict(name="extremes", lat=lat, lon=lon, elev=elev, conformal=False, aci=False,
                init_state=None, events=rec.events, final=rec.final())


def calendar_cases():
    """Часы от эпохи вокруг границ годов, включая 2000 (високосный) и 2100 (нет)."""
    from mayak.timeaxis import to_utc_hour
    pts = ["1970-01-01T00", "1999-12-31T23", "2000-02-28T12", "2000-12-31T23",
           "2023-12-31T20", "2024-02-28T22", "2024-12-31T23", "2025-06-15T11",
           "2100-02-28T23", "2100-12-31T23"]
    base = [int(to_utc_hour(np.datetime64(p, "s"))) for p in pts]
    hours = sorted({h + d for h in base for d in (-1, 0, 1, 2)})
    doy, hr = window_calendar(0, np.array(hours, np.int64))
    return dict(unix_hours=hours, doy=[_f(v) for v in doy], hour=[_f(v) for v in hr])


def calibration_cases(model, blob):
    """Калибровка как отдельный узел: calibrate_forecast и aci_run эталона."""
    rng = np.random.default_rng(GOLDEN_SEED)
    H = model.cfg.horizon
    cases = []
    for theta, conf in ((0.0, True), (0.37, True), (-0.52, True), (0.37, False)):
        q = np.sort(rng.normal(0.0, 3.0, (H, len(model.cfg.quantiles))), axis=-1)
        q[5, 2] = q[5, 4] + 0.5                          # немонотонный вход
        q = q.astype(np.float32)
        out, _mu = calibrate_forecast(q, GOLDEN_SHIFT if conf else None, _f(theta))
        cases.append(dict(q=blob.put(q), theta=_f(theta), conformal=conf,
                          expect=blob.put(out)))
    scores = rng.exponential(0.8, 400)
    scores[::37] = np.nan
    scores[5] = np.inf
    run = aci_run(scores, GOLDEN_ACI, theta0=0.1)
    aci = dict(scores=[_enc_score(v) for v in scores],
               theta0=0.1, theta_before=[_f(v) for v in run["theta"]],
               miss=[None if np.isnan(v) else bool(v) for v in run["miss"]],
               theta_end=_f(run["theta_end"]))
    ys = rng.normal(0.0, 4.0, 64)
    qs = np.sort(rng.normal(0.0, 3.0, (64, len(model.cfg.quantiles))), axis=-1).astype(np.float32)
    qs[3] = qs[3, 3]                                     # вырожденный интервал
    ys[7] = qs[7, 3]
    sc = aci_score(ys.astype(np.float32).astype(np.float64), qs, GOLDEN_ACI.interval)
    score = dict(y=[_f(v) for v in ys], q=blob.put(qs),
                 expect=[_enc_score(v) for v in sc])
    return dict(cases=cases, aci=aci, score=score)


def generate(out_dir=DEFAULT_DIR):
    """Полная генерация эталона → golden.json. Модель и графы - в out_dir/model."""
    from mayak.runtime.graphs import export_graphs
    os.makedirs(out_dir, exist_ok=True)
    model = golden_model()
    manifest = export_graphs(model, os.path.join(out_dir, "model"), conformal=GOLDEN_SHIFT,
                             aci=GOLDEN_ACI)
    manifest["provenance"].pop("created_utc", None)
    with open(os.path.join(out_dir, "model", "manifest.json"), "w", encoding="utf-8") as fh:
        json.dump(manifest, fh, ensure_ascii=False, indent=1)
    blob = _Blob()
    cold, m1, _H = scenario_cold_aci(model, blob)
    restart, m2 = scenario_restart(model, blob, out_dir)
    extremes = scenario_extremes(model, blob)
    margins = m1 + m2
    if margins and min(margins) < MIN_ACI_MARGIN:
        raise RuntimeError(f"обратная связь ACI эталона в {min(margins):.1e} от порога e^θ: "
                           f"решение о промахе неустойчиво к float32; смените сид")
    doc = dict(format=GOLDEN_FORMAT, seed=GOLDEN_SEED, year_hours=YEAR_H,
               tolerance=dict(q_abs=Q_ATOL, fresh_abs=FRESH_ATOL),
               aci_margin_min=float(min(margins)) if margins else None,
               scenarios=[cold, restart, extremes], calendar=calendar_cases(),
               calibration=calibration_cases(model, blob))
    blob.array().astype("<f4").tofile(os.path.join(out_dir, "golden.f32"))
    with open(os.path.join(out_dir, "golden.json"), "w", encoding="utf-8") as fh:
        json.dump(doc, fh, ensure_ascii=False, separators=(",", ":"))
    return doc


def load(out_dir=DEFAULT_DIR):
    with open(os.path.join(out_dir, "golden.json"), encoding="utf-8") as fh:
        doc = json.load(fh)
    blob = np.fromfile(os.path.join(out_dir, "golden.f32"), "<f4")
    return doc, blob


def take(blob, ref, shape=None):
    a = blob[ref["offset"]:ref["offset"] + ref["len"]]
    return a.reshape(shape) if shape is not None else a


def replay(doc, blob, runtime_factory, horizon):
    """Прогон сценариев эталона через любую реализацию → max|Δq| по сценариям.

    runtime_factory(scenario) → объект с step(T, P, RH, doy, hour) и forecast(doy, hour),
    возвращающим квантили (H, NQ) (или пару (q, mu)); для сценария с init_state фабрика
    сама загружает состояние.
    """
    out = {}
    for sc in doc["scenarios"]:
        rt = runtime_factory(sc)
        err = 0.0
        for ev in sc["events"]:
            if ev["op"] == "step":
                rt.step(*(dec_obs(v) for v in ev["obs"]), np.float32(ev["doy"]),
                        np.float32(ev["hour"]))
                continue
            q = rt.forecast(*_fut(ev["fut_hoy0"], horizon))
            if ev["q"] is None:
                continue
            q = q[0] if isinstance(q, tuple) else q
            ref = take(blob, ev["q"], (horizon, -1))
            err = max(err, float(np.abs(np.asarray(q) - ref).max()))
        out[sc["name"]] = err
    return out


__all__ = ["DEFAULT_DIR", "FRESH_ATOL", "GOLDEN_ACI", "GOLDEN_SHIFT", "Q_ATOL", "generate",
           "dec_obs", "golden_model", "load", "replay", "take"]
